File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        // check rows
        for(int i = 0; i < 9; i++) {
            set<char> s;
            for (int j = 0; j < 9; j++) {
                if(s.find(board[i][j]) != s.end()) {
                    cout<<board[i][j]<<" "<<endl;
                    return false;
                }
                if(isalnum(board[i][j])){
                    s.insert(board[i][j]);
                }
            }
        }

        // Check cols
        unordered_map<int, set<char>> m;
        for(int i = 0; i < 9; i++) {
            set<char> s;
            for (int j = 0; j < 9; j++) {
                if(s.find(board[j][i]) != s.end()) {
                    cout<<'here'<<endl;
                    return false;
                }
                if(isalnum(board[j][i])) {
                    s.insert(board[j][i]);
                }
            }
        }

        // check in each 3x3 box
        vector<vector<set<char>>> s;

        for(int i = 0; i < 3; i++) {
            vector<set<char>> temp;
            s.push_back(temp);
            for (int j = 0; j < 3; j++) {
                set<char> t;
                s[i].push_back(t);
            }
        }

        for(int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if(s[(int)i/3][(int)j/3].find(board[i][j]) != s[(int)i/3][(int)j/3].end()) {
                    cout<<board[i][j]<<" "<<i<<" "<<j<<endl;
                    return false;
                }
                if(isalnum(board[i][j])){
                    s[(int)i/3][(int)j/3].insert(board[i][j]);
                }
            }
        }

        return true;
    }
};

File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        # check rows and cols
        rows = defaultdict(set)
        cols = defaultdict(set)
        for r in range(len(board)):
            for c in range(len(board[r])):
                if (board[r][c] in rows[r] or board[r][c] in cols[c]) and board[r][c] != ".":
                    return False
                rows[r].add(board[r][c])
                cols[c].add(board[r][c])
        
        # check 3x3 boxes
        boxes = defaultdict(set)
        for R in range(0, len(board), 3):
            for C in range(0, len(board), 3):
                for r in range (R, R+3):
                    for c in range(C, C+3):
                        if board[r][c] in boxes[(R,C)] and board[r][c] != ".":
                            return False
                        if board[r][c] != ".":
                            boxes[(R,C)].add(board[r][c])
        print(boxes)

        return True